### ISSUE TITLE
torch.nonzero with known count to avoid host-device synchronization

### ISSUE THREAD

**Issue by albertz** [pytorch/pytorch#131256]
### 🚀 The feature, motivation and pitch

`torch.nonzero` causes a host-device synchronization on CUDA. This is because it must first calculate the count of nonzero elements, and then copy this count to host.

There are cases where I can easily calculate the count myself (from existing host data), or already know the count, which would make this unnecessary.

### Alternatives

For some given `mask` of shape `[in_len]`:
Instead of:
```python
idx = torch.nonzero(mask).squeeze(1)  # [out_len]
```
Do this trickery:
```python
idx = torch.argsort(mask.to(torch.int8), stable=True, descending=True)  # [in_len]
idx = idx[:out_len]  # [out_len]
```
This avoids the host-device synchronization, but on the other hand, is more complicated, and a bit slower.

I'm not sure if there are any other ways?

Btw, I'm using `torch.utils.benchmark.Timer` to measure the speed. But actually I'm not sure if this really properly covers the problem of host-device synchronization with `torch.nonzero`. Does it? E.g. when comparing this trickery code to `torch.nonzero`, I just see that `torch.argsort` is slower.

**Comment by eellison** [comment]
dupe of https://github.com/pytorch/pytorch/issues/62320 i think

**Comment by albertz** [comment]
> dupe of #62320 i think

Ah yes, I think you are right.
